mirgecom/euler.py

```diff
@@ -20,17 +22,193 @@
+        lambda el: discr.opposite_face_connection()(el.queue, el), i
+    )
+    return TracePair("int_faces", i, e)
+
+
+def _inviscid_flux(discr, q, eos=IdealSingleGas()):
+
+    ndim = discr.dim
+
+    # q = [ rho rhoE rhoV ]
+    rho = q[0]
+    rhoE = q[1]
+    rhoV = q[2:]
+
+    p = eos.Pressure(q)
+
+    def scalevec(scalar, vec):
+        # workaround for object array behavior
+        return make_obj_array([ni * scalar for ni in vec])
+    
+    # physical flux =
+    # [ rhoV (rhoE + p)V (rhoV.x.V + delta_ij*p) ]
+
+    momFlux = make_obj_array(
+        [
+            (rhoV[i] * rhoV[j] / rho + (p if i == j else 0))
+            for i in range(ndim)
+            for j in range(ndim)
+        ]
+    )
+
+    flux = join_fields(
+        scalevec(1.0, rhoV), scalevec((rhoE + p) / rho, rhoV), momFlux,
+    )
+
+    return flux
+
+
+def _get_wavespeeds(w_tpair, eos=IdealSingleGas()):
+
+    rho = w_tpair[0]
+    rhoE = w_tpair[1]
+    rhoV = w_tpair[2:]
+
+    def scalevec(scalar, vec):
+        # workaround for object array behavior
+        return make_obj_array([ni * scalar for ni in vec])
+
+    v_int, v_ext = [
+        scalevec(1.0 / lrho, lrhov)
+        for lrho, lrhov in [(rho.int, rhoV.int), (rho.ext, rhoV.ext)]
+    ]
+
+    qint = join_fields( rho.int, rhoE.int, rhoV.int)
+    qext = join_fields( rho.ext, rhoE.ext, rhoV.ext)
+    
+    pint, pext = [ eos.Pressure(qint), eos.Pressure(qext) ]
+    
+    c_int, c_ext = [ eos.SpeedOfSound(qint), eos.SpeedOfSound(qext) ]
+
+    fspeed_int, fspeed_ext = [
+        (clmath.sqrt(np.dot(lv, lv)) + lc)
+        for lv, lc in [(v_int, c_int), (v_ext, c_ext)]
+    ]
+
+    return join_fields(fspeed_int, fspeed_ext)
+
+
+def _facial_flux(discr, w_tpair, eos=IdealSingleGas()):
+
+    dim = discr.dim
+
+    rho = w_tpair[0]
+    rhoE = w_tpair[1]
+    rhoV = w_tpair[2:]
+
+    def scalevec(scalar, vec):
+        # workaround for object array behavior
+        return make_obj_array([ni * scalar for ni in vec])
+
+    normal = with_queue(rho.int.queue, discr.normal(w_tpair.dd))
+
+    # Get inviscid fluxes [rhoV (rhoE + p)V (rhoV.x.V + delta_ij*p) ]
+    qint = join_fields(rho.int, rhoE.int, rhoV.int)
+    qext = join_fields(rho.ext, rhoE.ext, rhoV.ext)
+    
+    # - Figure out how to manage grudge branch dependencies
+    #    qjump = join_fields(rho.jump, rhoE.jump, rhoV.jump)
+    qjump = qext - qint
+    flux_int = _inviscid_flux(discr, qint, eos)
+    flux_ext = _inviscid_flux(discr, qext, eos)
+
+    # Lax/Friedrichs/Rusonov after JSH/TW Nodal DG Methods, p. 209
+    #    flux_jump = scalevec(1.0,(flux_int - flux_ext))
+    flux_jump = scalevec(0.5, (flux_int + flux_ext))
+
+    # wavespeeds = [ wavespeed_int, wavespeed_ext ]
+    wavespeeds = _get_wavespeeds(w_tpair,eos)
+    
+    # - Gak!  What's the matter with this next line?   (CL issues?)
+    #    lam = np.maximum(fspeed_int,fspeed_ext)
+    # lam = 0.5*np.maximum(wavespeeds[0],wavespeeds[1])
```

Comment:
`pyopencl.array.maximum`

Reply:
Fixed!  Thanks so much!   I lost hair over this.

Reply:
> I lost hair over this.

It's up to you to dial in exactly how much hair to lose. :) It's totally OK to ask me stuff after you've been stuck on something for some period of time, where that period of time is definitely shorter than a week. (maybe  a half hour?)

That's also why these pull requests can be useful. You can just point at some code and be like, "@inducer, how do I fix this?" I like that better than Slack because it doesn't force me to respond right in that moment.